From df208a84aa1a260df329a8331840740e590b540a Mon Sep 17 00:00:00 2001
From: LibXZR <i@xzr.moe>
Date: Tue, 3 May 2022 22:44:21 +0530
Subject: [PATCH] base: Whitelist a few services to location indicator
 whitelist

commit 1f5d42679a4128c6f4948086d4567d67b68907e7
Author: Daniel Micay <danielmicay@gmail.com>
Date:   Tue May 3 09:23:38 2022 -0400

    add com.android.phone to indicator exemptions

    Phone services (not the Dialer) accesses detailed cellular information
    which is considered to be location information. It's a base OS component
    serving the intended purpose and shouldn't trigger spurious location
    indicator notices every time it retrieves cellular information.

commit f1fffae1f307fbdb0df71e265d8c3435199279b1
Author: Daniel Micay <danielmicay@gmail.com>
Date:   Sun Apr 17 14:45:50 2022 -0400

    add com.android.bluetooth to indicator exemptions

commit 3099b411895b64eec5ef77bd21a65141cbd092b4
Author: Chiranth A J <chiranth@m.ms.evolution-x.org>
Date:   Sun May 1 22:46:56 2022 +0530

    Add Mediatek IMS to location indicator whitelist

    Spams everytime on mediatek devices

commit b8da3479d7b126ec8e575fe75abb3373aa673f89
Author: Joey Huab <joey@evolution-x.org>
Date:   Tue May 3 18:32:46 2022 +0900

    Add Phone Services to location indicator whitelist

commit cd6445fce4bf853542974df820d7b98623ee8509
Author: Chiranth A J <chiranth@m.ms.evolution-x.org>
Date:   Sun Mar 6 20:22:35 2022 +0000

    Add Settings services to location indicator whitelist

    Signed-off-by: Chiranth A J <chiranth@m.ms.evolution-x.org>
    Change-Id: I6942e94611d4b70091d6553c26652b2eb743bc1e

commit 1f1500c62ed503311c2d45d0f677909d6a72eab3
Author: LibXZR <i@xzr.moe>
Date:   Sun Mar 6 11:05:24 2022 +0800

    Add Tethering and SystemUI to location indicator whitelist

    They are reported to be disturbing as well.

    Change-Id: I6776860e8d2c7331ee281d093badb102b083fdd9
    Signed-off-by: LibXZR <i@xzr.moe>

commit 2e1f076e01ec73dc2ef3897f186b7449404e5301
Author: LibXZR <i@xzr.moe>
Date:   Fri Mar 4 11:14:40 2022 +0800

    Do not show location indicator for bluetooth package

    The location indicator is frequently showed when wireless headsets
    are connected, which disturbes a lot especially when watching videos
    or playing games in landscape mode.

    Change-Id: I94deddb62b5b48d3eee153880d40c1928e776ebc
    Signed-off-by: LibXZR <i@xzr.moe>

Co-authored-by: Daniel Micay <danielmicay@gmail.com>
Co-authored-by: Chiranth A J <chiranth@m.ms.evolution-x.org>
Co-authored-by: Joey Huab <joey@evolution-x.org>
---
 .../android/permission/PermissionManager.java    |  4 ++++
 .../systemui/privacy/PrivacyItemController.kt    | 16 +++++++++++++---
 2 files changed, 17 insertions(+), 3 deletions(-)

diff --git a/core/java/android/permission/PermissionManager.java b/core/java/android/permission/PermissionManager.java
index 8a5f59a5cb6c..50cd3ad9f1b5 100644
--- a/core/java/android/permission/PermissionManager.java
+++ b/core/java/android/permission/PermissionManager.java
@@ -91,6 +91,8 @@ public final class PermissionManager {
             "permission grant or revoke changed gids";
 
     private static final String SYSTEM_PKG = "android";
+    private static final String BLUETOOTH_PKG = "com.android.bluetooth";
+    private static final String PHONE_SERVICES_PKG = "com.android.phone";
 
     /**
      * Refuse to install package if groups of permissions are bad
@@ -940,6 +942,8 @@ public static Set<String> getIndicatorExemptedPackages(@NonNull Context context)
         updateIndicatorExemptedPackages(context);
         ArraySet<String> pkgNames = new ArraySet<>();
         pkgNames.add(SYSTEM_PKG);
+        pkgNames.add(BLUETOOTH_PKG);
+        pkgNames.add(PHONE_SERVICES_PKG);
         for (int i = 0; i < INDICATOR_EXEMPTED_PACKAGES.length; i++) {
             String exemptedPackage = INDICATOR_EXEMPTED_PACKAGES[i];
             if (exemptedPackage != null) {
diff --git a/packages/SystemUI/src/com/android/systemui/privacy/PrivacyItemController.kt b/packages/SystemUI/src/com/android/systemui/privacy/PrivacyItemController.kt
index 25dee4d2088c..a3ebe3603776 100644
--- a/packages/SystemUI/src/com/android/systemui/privacy/PrivacyItemController.kt
+++ b/packages/SystemUI/src/com/android/systemui/privacy/PrivacyItemController.kt
@@ -57,6 +57,14 @@ class PrivacyItemController @Inject constructor(
 
     @VisibleForTesting
     internal companion object {
+        val LOCATION_WHITELIST_PKG = arrayOf(
+            "com.android.bluetooth",
+            "com.android.networkstack.tethering",
+            "com.android.phone",
+            "com.android.systemui",
+            "com.google.android.settings.intelligence",
+            "com.mediatek.ims",
+        )
         val CAMERA_WHITELIST_PKG = arrayOf(
             "org.pixelexperience.faceunlock",
         )
@@ -150,7 +158,8 @@ class PrivacyItemController @Inject constructor(
             active: Boolean
         ) {
             // Check if we care about this code right now
-            if (code in OPS_LOCATION && !locationAvailable) {
+            if (code in OPS_LOCATION && !locationAvailable
+                    || packageName in LOCATION_WHITELIST_PKG) {
                 return
             }
             if (code in OPS_MIC_CAMERA && !micCameraAvailable
@@ -326,7 +335,8 @@ class PrivacyItemController @Inject constructor(
             AppOpsManager.OP_RECORD_AUDIO -> PrivacyType.TYPE_MICROPHONE
             else -> return null
         }
-        if (type == PrivacyType.TYPE_LOCATION && !locationAvailable) {
+        if (type == PrivacyType.TYPE_LOCATION && !locationAvailable
+                || appOpItem.packageName in LOCATION_WHITELIST_PKG) {
             return null
         }
         if (type == PrivacyType.TYPE_CAMERA && !micCameraAvailable
@@ -392,4 +402,4 @@ class PrivacyItemController @Inject constructor(
             listeningCanceller = delegate.executeDelayed({ setListeningState() }, 0L)
         }
     }
-}
\ No newline at end of file
+}
